### ISSUE TITLE
Parsing GTFS time optimization ?

### ISSUE THREAD

**Issue by esdras-kid** [valhalla/valhalla#5003]
I'm trying to build a graph on all Germany with transit. 
As you may already know [gtfs.de ](https://gtfs.de/) furnish gtfs at country level for the hole country ,
the problem with this, I have been running the parsing of the gtfs for a 3 days now.

So i took a look on the code for ingets_transit and with some log i found out the part handling the  stop time take time for each stop :
Inside the select_transit_tiles function 
```cpp
for (const auto& stopTime : stopTimess) {
          // add trip, route, agency and service_id from stop_time, it's the only place with that info
          // TODO: should we throw here?
          auto trip = feed.get_trip(stopTime.trip_id);
          auto route = feed.get_route(trip.route_id);
          if (!gtfs::valid(trip) || !gtfs::valid(route) || trip.service_id.empty()) {
            LOG_ERROR("Missing trip or route or service_id for trip");
            continue;
          }

          tile_info.trips.insert({trip.trip_id, feed_name});
          tile_info.routes.insert({{route.route_id, feed_name}, tile_info.routes.size()});

          // shapes are optional, don't keep non-existing shapes around
          if (!trip.shape_id.empty()) {
            tile_info.shapes.insert({{trip.shape_id, feed_name}, tile_info.shapes.size()});
          }
}
```

What I'm proposing and coming for someone that never wrote a c++ code, I want to have your idea on it before I try it :

Parallelize the part on stoptimes by tile after the insert of all the stations_childrens:

```cpp

#include <thread>
#include <vector>
#include <future>
#include <mutex>

std::priority_queue<tile_transit_info_t> select_transit_tiles(const std::string& gtfs_path) {
  std::set<GraphId> tiles;
  const auto& local_tiles = TileHierarchy::levels().back().tiles;
  std::unordered_map<GraphId, tile_transit_info_t> tile_map;

  // Function to get tile info
  auto get_tile_info = [&tile_map, &local_tiles, &tile_map_mutex](const gtfs::Stop& stop) -> tile_transit_info_t& {
    auto graphid = GraphId(local_tiles.TileId(stop.stop_lat, stop.stop_lon),
                           TileHierarchy::GetTransitLevel().level, 0);
    return tile_map.insert({graphid, tile_transit_info_t{graphid}}).first->second;
  };

  filesystem::recursive_directory_iterator gtfs_feed_itr(gtfs_path);
  filesystem::recursive_directory_iterator end_file_itr;

  for (; gtfs_feed_itr != end_file_itr; ++gtfs_feed_itr) {
    const auto& feed_path = gtfs_feed_itr->path();
    if (gtfs_feed_itr->is_directory() && filesystem::is_empty(feed_path)) {
      LOG_ERROR("Feed directory " + feed_path.string() + " is empty");
      continue;
    }

    if (filesystem::is_directory(feed_path)) {
      const auto feed_name = feed_path.filename().string();
      LOG_INFO("Loading " + feed_name);

      // Load the feed
      gtfs::Feed feed(feed_path.string());
      auto read_result = feed.read_feed();
      if (read_result.code != gtfs::ResultCode::OK) {
        LOG_ERROR("Couldn't find a required file for feed " + feed_path.filename().string() + ": " +
                  read_result.message);
        continue;
      }

      const auto& stops = feed.get_stops();
      LOG_INFO("Start processing stops for feed " + feed_name);

      // 1st pass: Add all stations
      for (const auto& stop : stops) {
        if (stop.location_type == gtfs::StopLocationType::Station) {
          auto& tile_info = get_tile_info(stop);
          tile_info.stations.insert({stop.stop_id, feed_name});
        }
      }

      // 2nd pass: Add platforms/stops
      for (const auto& stop : stops) {
        if (!(stop.location_type == gtfs::StopLocationType::StopOrPlatform) &&
            !(stop.location_type == gtfs::StopLocationType::EntranceExit)) {
          continue;
        }

        auto& tile_info = get_tile_info(stop);

        const auto& parent_station = feed.get_stop(stop.parent_station);
        if (gtfs::valid(parent_station)) {
          tile_info.station_children.insert({{stop.parent_station, feed_name}, stop.stop_id});
        } else {
          tile_info.stations.insert({stop.stop_id, feed_name});
          tile_info.station_children.insert({{stop.stop_id, feed_name}, stop.stop_id});
        }
      }

      LOG_INFO("Done processing stops for GTFS feed " + feed_name);

      // Now process stop times for this feed in parallel by graphid
      LOG_INFO("Starting stop times processing for " + feed_name + " in parallel");

      std::vector<std::future<void>> futures;

      // Process stop times for each station in station_children in parallel
      for (auto& [graphid, tile_info] : tile_map) {
        // If the graphid belongs to the current feed, process it in parallel
        futures.push_back(std::async(std::launch::async, [&]() {
          for (const auto& station_child : tile_info.station_children) {
            if (station_child.first.second != feed_name) {
              continue; // Skip children stations not from the current feed
            }

            const auto& stop_id = station_child.second;
            const auto& stopTimess = feed.get_stop_times_for_stop(stop_id);

            for (const auto& stopTime : stopTimess) {
              auto trip = feed.get_trip(stopTime.trip_id);
              auto route = feed.get_route(trip.route_id);
              if (!gtfs::valid(trip) || !gtfs::valid(route) || trip.service_id.empty()) {
                LOG_ERROR("Missing trip or route or service_id for trip");
                continue;
              }

              std::lock_guard<std::mutex> lock(tile_map_mutex);
              tile_info.trips.insert({trip.trip_id, feed_name});
              tile_info.routes.insert({{route.route_id, feed_name}, tile_info.routes.size()});
              if (!trip.shape_id.empty()) {
                tile_info.shapes.insert({{trip.shape_id, feed_name}, tile_info.shapes.size()});
              }
            }
          }
        }));
      }

      // Wait for all stop times processing tasks to finish
      for (auto& future : futures) {
        future.get();
      }

      LOG_INFO("Done stop times processing for feed " + feed_name);
    }
  }

  // Finally, construct the priority queue of tile info
  std::priority_queue<tile_transit_info_t> queue;
  for (auto it = tile_map.begin(); it != tile_map.end(); it++) {
    queue.push(it->second);
  }

  return queue;
}


```

Thank you for your help

**Comment by CommanderStorm** [comment]
Your code is somewhat hard to parse (the reason you did not recive better feedback in the last month).
Could you submit code-suggestions via a pull request instead?

This way it is more clear what changes are done.

**Comment by esdras-kid** [comment]
Hello @CommanderStorm ,

I don't have access to this repository for pr, so i fork it and a pr on my account :
[Here is the branch](https://github.com/esdras-kid/valhalla/tree/feat/_test_opti_ingest_transit)

I try to run the code on my computer mac m2 and i can't run it on local without dockerizing.
Hence i'm not sure about my optimisation

**Comment by CommanderStorm** [comment]
> I don't have access to this repository for pr,

Shure.. you can create PRs..

https://github.com/valhalla/valhalla/compare/master...esdras-kid:valhalla:feat/_test_opti_ingest_transit

**Comment by kevinkreiser** [comment]
ah i see you said:

>  I want to have your idea on it before I try it 

the first thing i would say is that you'll notice that we have a bunch of maps trips and routes etc.  you'd need to synchronize access to these between threads. synchronization is a performance killer of course so better to avoid it.  if we parallelize on tiles then we can be sure no given thread will access the same tile at the same time. what we can do is pre allocate slots in an external map (a map of maps) where the key is the tile id and then the map inside it is what it looks like in the current code. y ou'll also see that we increment an index essentially globally which is based off of the containers size. this wouldnt work but instead we could keep an atomic counter and let the threads all access that without having to worry about synchronization.

so in short, this is not ready for a pr though you could pr it anyway as an idea and iterate on it there.

**Comment by esdras-kid** [comment]
Hey @CommanderStorm, thanks for your help! Here's the PR:
https://github.com/valhalla/valhalla/pull/5134

Yep @kevinkreiser, after some testing, I realized I was on the wrong track, so I decided to go with a different approach in this PR.

Instead of making multiple calls to feed.get_stop_times_for_stop(stop.stop_id), which can be slow for large GTFS data, I now loop through the stop times once and cache the trip ID and route ID.

I've also built a graph for France with this method, and I'll try it for Germany next (where I previously couldn't build it) and provide my feedback.
